include/RAJA/policy/hip/intrinsics.hpp

```diff
@@ -0,0 +1,362 @@
+ *
+ * \Note This uses device scope fences to ensure ordering and to flush local
+ *       caches so that memory accesses become visible to the whole device.
+ * \Note This class uses normal memory accesses that are cached in local caches
+ *       so device scope fences are required to make memory accesses visible
+ *       to the whole device.
+ */
+struct AccessorDeviceScopeUseLocalCache : RAJA::detail::DefaultAccessor
+{
+  static RAJA_DEVICE RAJA_INLINE void fence_acquire()
+  {
+    __threadfence();
+  }
+
+  static RAJA_DEVICE RAJA_INLINE void fence_release()
+  {
+    __threadfence();
+  }
+};
+
+/*!
+ ******************************************************************************
+ *
+ * \brief Abstracts access to memory when coordinating between threads at
+ *       device scope. The fences provided here are to be used with relaxed
+ *       atomics in order to guarantee memory ordering and visibility of the
+ *       accesses done through this class.
+ *
+ * \Note This may use block scope fences to ensure ordering and avoid flushing
+ *       local caches so special memory accesses are used to ensure visibility
+ *       to the whole device.
+ * \Note This class uses device scope atomic memory accesses to bypass local
+ *       caches so memory accesses are visible to the whole device without
+ *       device scope fences.
+ * \Note A memory access may be split into multiple memory accesses, so
+ *       even though atomic instructions are used concurrent accesses between
+ *       different threads are not thread safe.
+ *
+ ******************************************************************************
+ */
+struct AccessorDeviceScopeUseSharedCache
+{
+  // hip has 32 and 64 bit atomics
+  static constexpr size_t min_atomic_int_type_size = sizeof(unsigned int);
+  static constexpr size_t max_atomic_int_type_size = sizeof(unsigned long long);
+
+  template < typename T >
+  static RAJA_DEVICE RAJA_INLINE T get(T* in_ptr, size_t idx)
+  {
+    using ArrayType = RAJA::detail::AsIntegerArray<T, min_atomic_int_type_size, max_atomic_int_type_size>;
+    using integer_type = typename ArrayType::integer_type;
+
+    ArrayType u;
+    auto ptr = const_cast<integer_type*>(reinterpret_cast<const integer_type*>(in_ptr + idx));
+
+    for (size_t i = 0; i < u.array_size(); ++i) {
+#if defined(RAJA_USE_HIP_INTRINSICS) && RAJA_INTERNAL_CLANG_HAS_BUILTIN(__hip_atomic_load)
+      u.array[i] = __hip_atomic_load(&ptr[i], __ATOMIC_RELAXED, __HIP_MEMORY_SCOPE_AGENT);
+#else
+      u.array[i] = atomicAdd(&ptr[i], integer_type(0));
+#endif
+    }
+
+    return u.get_value();
+  }
+
+  template < typename T >
+  static RAJA_DEVICE RAJA_INLINE void set(T* in_ptr, size_t idx, T val)
+  {
+    using ArrayType = RAJA::detail::AsIntegerArray<T, min_atomic_int_type_size, max_atomic_int_type_size>;
+    using integer_type = typename ArrayType::integer_type;
+
+    ArrayType u;
+    u.set_value(val);
+    auto ptr = reinterpret_cast<integer_type*>(in_ptr + idx);
```

Comment:
I guess the HIP atomic intrinsics can only operate on integers?

Reply:
If I didn't do this I would have to have a case for the supported types, here I can use the integer case for everything.